utils.go

```diff
@@ -0,0 +1,160 @@
+package loop
+
+import (
+	"context"
+
+	"github.com/btcsuite/btcd/btcec"
+	"github.com/btcsuite/btcutil"
+	"github.com/lightninglabs/lndclient"
+	"github.com/lightningnetwork/lnd/lnwire"
+	"github.com/lightningnetwork/lnd/routing/route"
+	"github.com/lightningnetwork/lnd/zpay32"
+)
+
+var (
+	// DefaultMaxHopHints is set to 20 as that is the default set in LND
+	DefaultMaxHopHints = 20
+)
+
+// SelectHopHints is a direct port of the SelectHopHints found in lnd. It was
+// reimplemented because the current implementation in LND relies on internals
+// not externalized through the API. Hopefully in the future SelectHopHints
+// will be refactored to allow for custom data sources. It iterates through all
+// the active and public channels available and returns eligible channels.
+// Eligibility requirements are simple: does the channel have enough liquidity
+// to fulfill the request and is the node whitelisted (if specified)
+func SelectHopHints(ctx context.Context, lnd *lndclient.LndServices,
+	amtMSat btcutil.Amount, numMaxHophints int,
+	includeNodes map[route.Vertex]struct{}) ([][]zpay32.HopHint, error) {
+
+	// Fetch all active and public channels.
+	openChannels, err := lnd.Client.ListChannels(ctx, false, false)
+	if err != nil {
+		return nil, err
+	}
+
+	// We'll add our hop hints in two passes, first we'll add all channels
+	// that are eligible to be hop hints, and also have a local balance
+	// above the payment amount.
+	var totalHintBandwidth btcutil.Amount
+	hopHintChans := make(map[int]*lndclient.ChannelEdge)
+	hopHints := make([][]zpay32.HopHint, 0, numMaxHophints)
+
+	for i, channel := range openChannels {
+		// In this first pass, we'll ignore all channels in
+		// isolation can't satisfy this payment.
+
+		// Retrieve extra info for each channel not available in listChannels
+		chanInfo, err := lnd.Client.GetChanInfo(ctx, channel.ChannelID)
+		if err != nil {
+			return nil, err
+		}
+		// Cache the GetChanInfo result since it might be useful
+		hopHintChans[i] = chanInfo
+		if channel.RemoteBalance < amtMSat {
+			continue
+		}
+		// If includeNodes is set, we'll only add channels with peers in
+		// include_node. This is done to respect the last_hop parameter
+		if len(includeNodes) > 0 {
+			if _, ok := includeNodes[channel.PubKeyBytes]; !ok {
+				continue
+			}
+		}
+
+		// We want to prevent leaking private nodes, which we define as
+		// nodes with only private channels
+		//
+		// GetNodeInfo will never return private channels, even if they're
+		// somehow known to us. If there are any channels returned, we can
+		// consider the node to be public.
+		nodeChannels, err := lnd.Client.GetNodeInfo(
+			ctx, channel.PubKeyBytes, true,
+		)
+		if err != nil {
```

Comment:
> If we have a [private/nonprivate] channel with a peer, we'll have it in our graph

Are you certain this is true? Sometimes lnd's graph can lag a bit, so I think it's possible that we have a channel with a peer but they're not yet in our graph. Don't want the swap to error out just because we don't know about a channel yet, when we could just skip it?